components/sup/src/manager/service/spec.rs

```diff
@@ -337,27 +337,94 @@ impl ServiceSpec {
                     // have to do anything.
                     None
                 } else {
-                    // TODO (CM): In the future, this would be the
-                    // place where we can evaluate what has changed
-                    // between the spec-on-disk and our in-memory
-                    // representation and potentially just bring our
-                    // in-memory representation in line without having
-                    // to restart the entire service.
-                    debug!("Reconciliation: '{}' queued for restart",
-                           running_spec.ident);
-                    Some(ServiceOperation::Restart { to_stop:  running_spec,
-                                                     to_start: disk_spec, })
+                    // Currently, if any of these bits of data are
+                    // different, we should restart the service. This
+                    // is not to say that it will *always* be that
+                    // way, however. Initially we are allowing dynamic
+                    // update of update-related configuration, but
+                    // nothing prevents us from making more things
+                    // dynamic in the future. We are proceeding
+                    // conservatively.
+
+                    // NOTE: There is no need to compare desired
+                    // state, because we know they must both be `Up`
+                    // for us to be in this code.
+
+                    // NOTE: if the idents change in any way, you
+                    // *must* restart, since that change may result in
+                    // a different version of the service being run.
+                    if running_spec.ident != disk_spec.ident
+                        || running_spec.group != disk_spec.group
+                        // TODO (CM): This *might* not need to be here
+                        || running_spec.topology != disk_spec.topology
+                        // TODO (CM): Bind information *may* be able
+                        // to be dynamically changed, but that will
+                        // need to be investigated more deeply.
+                        || running_spec.binds != disk_spec.binds
+                        || running_spec.binding_mode != disk_spec.binding_mode
+                        || running_spec.config_from != disk_spec.config_from
+                        // TODO (CM): This probably doesn't need to be here
+                        || running_spec.shutdown_timeout != disk_spec.shutdown_timeout
+                        || running_spec.svc_encrypted_password != disk_spec.svc_encrypted_password
+                        // TODO (CM): This probably doesn't need to be here, either
+                        || running_spec.health_check_interval != disk_spec.health_check_interval
+                    {
```

Comment:
Could we use pattern matching to pull out the fields on one (or both?) of the `ServiceSpec`s here? If we explicitly list every field in the pattern match, if we add a new field to `ServiceSpec` the compiler will automatically pull us to this section of the code. This will force us to evaluate how the update behavior of the field should work.

Reply:
Do you have a concrete solution for that? I agree that it's a bummer that this isn't as compiler-driven as I would like, for the reasons you mention. I couldn't come up with a way to do that, though. If you've got a way to do it that I overlooked, that would be great to add!

Reply:
It is not perfect, but just pattern marching on all the fields of `ServiceSpec` will ensure we examine the update logic should a new field be added.

```rust
                    // Explicitly destructure every field here. This will force us to examine the
                    // update logic for any new fields added to `ServiceSpec`.
                    let ServiceSpec {
                        ident,
                        group,
                        bldr_url,
                        channel,
                        topology,
                        update_strategy,
                        update_condition,
                        binds,
                        binding_mode,
                        config_from,
                        desired_state: _,
                        shutdown_timeout,
                        svc_encrypted_password,
                        health_check_interval,
                    } = &running_spec;

                    if ident != &disk_spec.ident ...
```

Reply:
Ah, I see what you mean... yes, we can do that for sure :+1: